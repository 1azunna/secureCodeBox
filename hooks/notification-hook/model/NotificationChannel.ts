import { NotifierType } from "../NotifierType"

export interface NotificationChannel {
  name: string;
  type: NotifierType;
  template: string;
  rules: any;
  endPoint: string;
}
