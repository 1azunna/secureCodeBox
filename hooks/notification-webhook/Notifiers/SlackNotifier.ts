import { AbstractNotifier } from "./AbstractNotifier"

export class SlackNotifier extends AbstractNotifier {
  constructor(template: string) {
    super(template)
  }
  loadDefaultTemplate() {
    this.template = this.template
  }
  sendNotification() {

  }
}
