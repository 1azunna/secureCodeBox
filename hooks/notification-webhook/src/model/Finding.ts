export interface Finding {
  name: string;
  location: string;
  category: string;
  severity: string;
  osi_layer: string;
  attributes: Map<string, string | number>;
}
