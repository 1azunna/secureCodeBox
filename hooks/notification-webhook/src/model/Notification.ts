import { NotifierType } from "../NotifierType"

export class Notification {
  public type: NotifierType
  public template: string
  public rules: any
}
