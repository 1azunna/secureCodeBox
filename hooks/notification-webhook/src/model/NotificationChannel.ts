import { NotifierType } from "../NotifierType"

export interface NotificationChannel {
  name: string;
  type: NotifierType;
  templateName: string;
  rules: any;
  endPoint: string;
}
