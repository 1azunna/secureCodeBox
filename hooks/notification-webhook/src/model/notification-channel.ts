import { NotifierType } from "../NotifierType"

export class NotificationChannel {
  public name: string;
  public type: NotifierType;
  public templateName: string;
  public rules: any;
  public endPoint: string;
}
