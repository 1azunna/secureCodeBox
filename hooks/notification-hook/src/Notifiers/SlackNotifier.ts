import { NotifierType } from "../NotifierType"
import { AbstractNotifier } from "./AbstractNotifier"
import { Finding } from "../model/Finding"
import axios from 'axios';
import { NotificationChannel } from "../model/NotificationChannel";
import { Scan } from "../model/Scan";

export class SlackNotifier extends AbstractNotifier {

  protected type: NotifierType = NotifierType.SLACK

  constructor(channel: NotificationChannel, scan: Scan, findings: Finding[], args: any[]) {
    super(channel, scan, findings, args);
  }

  public async sendMessage(): Promise<void> {
    await this.sendPostRequest(this.renderMessage());
  }

  protected async sendPostRequest(message: string) {
    try {
      await axios.post(this.channel.endPoint, message)
    } catch (e) {
      console.log(`There was an Error sending the Message for the Slack Notifier "${this.channel.name}"`);
      console.log(e);
    }
  }
}
