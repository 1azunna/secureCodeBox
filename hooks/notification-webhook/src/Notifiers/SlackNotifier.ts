import { NotifierType } from "../NotifierType"
import { AbstractNotifier } from "./AbstractNotifier"
import { Finding } from "../model/Finding"
import * as Mustache from 'mustache';
import { TemplateType } from "../templateType";

export class SlackNotifier extends AbstractNotifier {
  protected type: NotifierType = NotifierType.SLACK
  constructor() {
    super();
  }

  public async sendMessage(findings: Finding[]): Promise<void> {
    console.log("Render Slack Message");
    let template = Mustache.render(this.template, { scanner: "nmap" });
    await this.sendPostRequest(template);
  }


  public async initTemplate(templateName: string): Promise<void> {
    switch (templateName) {
      case TemplateType.MESSAGE_CARD:
        await this.loadTemplate(templateName)
      default:
        throw new Error(`The Template Type ${templateName} does not exist :(`)
    }
  }

  private async sendPostRequest(template: string) {
    console.log(`Template: ${template}`);
  }
}
