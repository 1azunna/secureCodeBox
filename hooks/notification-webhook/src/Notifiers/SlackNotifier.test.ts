import { SlackNotifier } from "./SlackNotifier";
import axios from 'axios'
import { NotificationChannel } from "../model/NotificationChannel";
import { NotifierType } from "../NotifierType";
import { Scan } from "../model/Scan";

jest.mock('axios');

test.only("Should Send Message With Findings And Severities", async () => {
  const message = `{\"blocks\":[{\"type\":\"header\",\"text\":{\"type\":\"plain_text\",\"text\":\"New Nmap security scan results are available!\",\"emoji\":true}},{\"type\":\"context\",\"elements\":[{\"type\":\"image\",\"image_url\":\"https://www.securecodebox.io/favicon.png\",\"alt_text\":\"secureCodeBox Favicon\"},{\"type\":\"mrkdwn\",\"text\":\"Scan: demo-scan-1601086432\\nCreated at Fri Jan 01 2021 15:29:25 GMT+0100 (Central European Standard Time)\\n\"}]},{\"type\":\"section\",\"fields\":[{\"type\":\"mrkdwn\",\"text\":\"\\n  *Findings Severity Overview*:\\n high: 10
medium: 5
low: 2
informational: 1
\\n\\n\"}]},{\"type\":\"section\",\"fields\":[{\"type\":\"mrkdwn\",\"text\":\"\\n  *Findings Category Overview*:\\n A Client Error response code was returned by the server: 1
Information Disclosure - Sensitive Information in URL: 1
Strict-Transport-Security Header Not Set: 1
\\n\\n\"}]},{\"type\":\"actions\",\"elements\":[{\"type\":\"button\",\"text\":{\"type\":\"plain_text\",\"text\":\"Open Kibana Dashboard\",\"emoji\":true},\"value\":\"click_me_123\",\"action_id\":\"actionId-0\",\"url\":\"https://securecodebox.io\"},{\"type\":\"button\",\"text\":{\"type\":\"plain_text\",\"text\":\"Show Results in Kibana Dashboard\",\"emoji\":true},\"value\":\"click_me_123\",\"action_id\":\"actionId-1\",\"url\":\"https://securecodebox.io\"}]}]}`

  const link = "https://hooks.slack.com/services/<YOUR_TOKEN>"

  const scan: Scan = {
    metadata: {
      uid: "09988cdf-1fc7-4f85-95ee-1b1d65dbc7cc",
      name: "demo-scan-1601086432",
      namespace: "my-scans",
      creationTimestamp: new Date("2021-01-01T14:29:25Z"),
      labels: {
        company: "iteratec",
        "attack-surface": "external",
      },
    },
    spec: {
      scanType: "Nmap",
      parameters: ["-Pn", "localhost"],
    },
    status: {
      findingDownloadLink:
        "https://my-secureCodeBox-instance.com/scan-b9as-sdweref--sadf-asdfsdf-dasdgf-asdffdsfa7/findings.json",
      findings: {
        categories: {
          "A Client Error response code was returned by the server": 1,
          "Information Disclosure - Sensitive Information in URL": 1,
          "Strict-Transport-Security Header Not Set": 1,
        },
        count: 3,
        severities: {
          high: 10,
          medium: 5,
          low: 2,
          informational: 1,
        },
      },
      finishedAt: new Date("2020-05-25T02:38:13Z"),
      rawResultDownloadLink:
        "https://my-secureCodeBox-instance.com/scan-blkfsdg-sdgfsfgd-sfg-sdfg-dfsg-gfs98-e8af2172caa7/zap-results.json?Expires=1601691232",
      rawResultFile: "zap-results.json",
      rawResultType: "zap-json",
      state: "Done",
    },
  };
  const channel: NotificationChannel = {
    name: "Channel Name",
    type: NotifierType.SLACK,
    templateName: "messageCard",
    rules: [],
    endPoint: "https://hooks.slack.com/services/<YOUR_TOKEN>"
  };
  const slackNotifier = new SlackNotifier(channel, scan);
  slackNotifier.sendMessage([]);
  expect(axios.post).toHaveBeenCalledWith(link, message);
});

test("Should Send Post Request with Findings", async () => {
  expect(true).toBe(false)
})

test("Should Send Minimal Template For Empty Findings", async () => {
  expect(true).toBe(false)
})

test("Should Include Link To Kibana Dashboard", async () => {
  expect(true).toBe(false)
})

test("Should Include Link To DefectDojo Project", async () => {
  expect(true).toBe(false)
})

test("Should Only Send Post Request for Matching Rules", async () => {
  expect(true).toBe(false)
})

test("Should Apply Custom Template", async () => {
  expect(true).toBe(false)
})

test("Should Send Minimal Template For Disabled VulnerabilityManagement", async () => {
  expect(true).toBe(false)
})

test("Should Send Minimal Template For Incomplete VulnerabilityManagement Configuration", async () => {
  expect(true).toBe(false)
})

test("Should Only Send Post Request For High Severity Findings", async () => {
  expect(true).toBe(false)
})
