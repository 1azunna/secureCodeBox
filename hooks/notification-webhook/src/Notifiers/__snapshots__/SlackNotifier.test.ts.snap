// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should Send Message With Findings And Severities 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "https://hooks.slack.com/services/<YOUR_TOKEN>",
      "{\\"blocks\\":[{\\"type\\":\\"header\\",\\"text\\":{\\"type\\":\\"plain_text\\",\\"text\\":\\"New Nmap security scan results are available!\\",\\"emoji\\":true}},{\\"type\\":\\"context\\",\\"elements\\":[{\\"type\\":\\"image\\",\\"image_url\\":\\"https://www.securecodebox.io/favicon.png\\",\\"alt_text\\":\\"secureCodeBox Favicon\\"},{\\"type\\":\\"mrkdwn\\",\\"text\\":\\"Scan: demo-scan-1601086432\\\\nCreated at Fri Jan 01 2021 15:29:25 GMT+0100 (Central European Standard Time)\\\\n\\"}]},{\\"type\\":\\"section\\",\\"fields\\":[{\\"type\\":\\"mrkdwn\\",\\"text\\":\\"\\\\n  *Findings Severity Overview*:\\\\n high: 10
medium: 5
low: 2
informational: 1
\\\\n\\\\n\\\\n\\"}]},{\\"type\\":\\"section\\",\\"fields\\":[{\\"type\\":\\"mrkdwn\\",\\"text\\":\\"\\\\n  *Findings Category Overview*:\\\\n A Client Error response code was returned by the server: 1
Information Disclosure - Sensitive Information in URL: 1
Strict-Transport-Security Header Not Set: 1
\\\\n\\\\n\\"}]},{\\"type\\":\\"actions\\",\\"elements\\":[{\\"type\\":\\"button\\",\\"text\\":{\\"type\\":\\"plain_text\\",\\"text\\":\\"Open Kibana Dashboard\\",\\"emoji\\":true},\\"value\\":\\"click_me_123\\",\\"action_id\\":\\"actionId-0\\",\\"url\\":\\"https://securecodebox.io\\"},{\\"type\\":\\"button\\",\\"text\\":{\\"type\\":\\"plain_text\\",\\"text\\":\\"Show Results in Kibana Dashboard\\",\\"emoji\\":true},\\"value\\":\\"click_me_123\\",\\"action_id\\":\\"actionId-1\\",\\"url\\":\\"https://securecodebox.io\\"}]}]}",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`Should Send Minimal Template For Empty Findings 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "https://hooks.slack.com/services/<YOUR_TOKEN>",
      "{\\"blocks\\":[{\\"type\\":\\"header\\",\\"text\\":{\\"type\\":\\"plain_text\\",\\"text\\":\\"New Nmap security scan results are available!\\",\\"emoji\\":true}},{\\"type\\":\\"context\\",\\"elements\\":[{\\"type\\":\\"image\\",\\"image_url\\":\\"https://www.securecodebox.io/favicon.png\\",\\"alt_text\\":\\"secureCodeBox Favicon\\"},{\\"type\\":\\"mrkdwn\\",\\"text\\":\\"Scan: demo-scan-1601086432\\\\nCreated at Fri Jan 01 2021 15:29:25 GMT+0100 (Central European Standard Time)\\\\n\\"}]},{\\"type\\":\\"section\\",\\"fields\\":[{\\"type\\":\\"mrkdwn\\",\\"text\\":\\"\\\\n  *Findings Severity Overview*:\\\\n high: 10
medium: 5
low: 2
informational: 1
\\\\n\\\\n\\\\n\\"}]},{\\"type\\":\\"section\\",\\"fields\\":[{\\"type\\":\\"mrkdwn\\",\\"text\\":\\"\\\\n  *Findings Category Overview*:\\\\n A Client Error response code was returned by the server: 1
Information Disclosure - Sensitive Information in URL: 1
Strict-Transport-Security Header Not Set: 1
\\\\n\\\\n\\"}]},{\\"type\\":\\"actions\\",\\"elements\\":[{\\"type\\":\\"button\\",\\"text\\":{\\"type\\":\\"plain_text\\",\\"text\\":\\"Open Kibana Dashboard\\",\\"emoji\\":true},\\"value\\":\\"click_me_123\\",\\"action_id\\":\\"actionId-0\\",\\"url\\":\\"https://securecodebox.io\\"},{\\"type\\":\\"button\\",\\"text\\":{\\"type\\":\\"plain_text\\",\\"text\\":\\"Show Results in Kibana Dashboard\\",\\"emoji\\":true},\\"value\\":\\"click_me_123\\",\\"action_id\\":\\"actionId-1\\",\\"url\\":\\"https://securecodebox.io\\"}]}]}",
    ],
    Array [
      "https://hooks.slack.com/services/<YOUR_TOKEN>",
      "{\\"blocks\\":[{\\"type\\":\\"header\\",\\"text\\":{\\"type\\":\\"plain_text\\",\\"text\\":\\"New Nmap security scan results are available!\\",\\"emoji\\":true}},{\\"type\\":\\"context\\",\\"elements\\":[{\\"type\\":\\"image\\",\\"image_url\\":\\"https://www.securecodebox.io/favicon.png\\",\\"alt_text\\":\\"secureCodeBox Favicon\\"},{\\"type\\":\\"mrkdwn\\",\\"text\\":\\"Scan: demo-scan-1601086432\\\\nCreated at Fri Jan 01 2021 15:29:25 GMT+0100 (Central European Standard Time)\\\\n\\"}]},{\\"type\\":\\"section\\",\\"fields\\":[{\\"type\\":\\"mrkdwn\\",\\"text\\":\\"\\\\n\\\\n  There are no findings to report.\\\\n\\\\n\\"}]},{\\"type\\":\\"section\\",\\"fields\\":[{\\"type\\":\\"mrkdwn\\",\\"text\\":\\"\\\\n\\"}]},{\\"type\\":\\"actions\\",\\"elements\\":[{\\"type\\":\\"button\\",\\"text\\":{\\"type\\":\\"plain_text\\",\\"text\\":\\"Open Kibana Dashboard\\",\\"emoji\\":true},\\"value\\":\\"click_me_123\\",\\"action_id\\":\\"actionId-0\\",\\"url\\":\\"https://securecodebox.io\\"},{\\"type\\":\\"button\\",\\"text\\":{\\"type\\":\\"plain_text\\",\\"text\\":\\"Show Results in Kibana Dashboard\\",\\"emoji\\":true},\\"value\\":\\"click_me_123\\",\\"action_id\\":\\"actionId-1\\",\\"url\\":\\"https://securecodebox.io\\"}]}]}",
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
